packages/runtime-core/src/componentOptions.ts

```diff
@@ -860,22 +860,14 @@ export function createWatcher(
         ? currentInstance
         : null
 
-    const newValue = getter()
-    if (
-      isArray(newValue) &&
-      isCompatEnabled(DeprecationTypes.WATCH_ARRAY, instance)
-    ) {
-      options.deep = true
```

Comment:
This check is redundant, there's no reason to set the option when the getter includes the compat implementation already. On top of that, doing the ahead-of-time check like that leads to incompatible behavior in the edge case of the first value being an array, but subsequent values not being arrays. 

If the compatibility goal is to make it so switching from `WATCH_ARRAY` compat to `deep: 1` will behave identically, having it set it statically *may* have merit, but the entire check seems unnecessary and breaks compatibility with Vue 2 behavior.

Reply:
Wait the tests are failing because of that? I'll take a look

Reply:
After more investigation it seems that the entire `traverse(val, 1)` block in the watch callback does nothing because of [this condition in watch.ts](https://github.com/vuejs/core/blob/main/packages/reactivity/src/watch.ts#L244-L248)

The issue seems to be that when `deep` is false, that code goes "okay cool, but has the value changed", while when `deep` is true it goes "okay bro you triggered a reaction, I'll trust you that *something* changed"

That would be fine, we could just set `deep = 1`, however that also means that this compat *only works* when the first value it retrieves is an array. If the initial value is `null` and later it returns an array, it won't work.

I came up with one solution that, while not great, does work. We make the compat return `{WATCH_ARRAY_UNWRAP: val}` on arrays, then add some special casing inside that if block that unwraps those special values. Because it's a new object every time it'll return true from the `hasChanged`, then get unwrapped to the underlying array later.

It's getting late, I'll tidy it up and commit tomorrow.